win32ss/user/user32/controls/ghost.c

```diff
@@ -0,0 +1,570 @@
+
+#include <user32.h>
+
+WINE_DEFAULT_DEBUG_CHANNEL(ghost);
+
+#define GHOST_TIMER_ID  0xFACEDEAD
+#define GHOST_INTERVAL  1000        // one second
+
+static const WCHAR ghostW[] = L"Ghost";
+const struct builtin_class_descr GHOST_builtin_class =
+{
+    ghostW,                     /* name */
+    0,                          /* style  */
+    GhostWndProcA,              /* procA */
+    GhostWndProcW,              /* procW */
+    0,                          /* extra */
+    IDC_WAIT,                   /* cursor */
+    (HBRUSH)(COLOR_3DFACE + 1)  /* brush */
+};
+
+static HBITMAP
+IntCreate32BppBitmap(INT cx, INT cy)
+{
+    HBITMAP hbm = NULL;
+    BITMAPINFO bi;
+    HDC hdc;
+    LPVOID pvBits;
+
+    ZeroMemory(&bi, sizeof(bi));
+    bi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+    bi.bmiHeader.biWidth = cx;
+    bi.bmiHeader.biHeight = cy;
+    bi.bmiHeader.biPlanes = 1;
+    bi.bmiHeader.biBitCount = 32;
+
+    hdc = CreateCompatibleDC(NULL);
+    if (hdc)
+    {
+        hbm = CreateDIBSection(hdc, &bi, DIB_RGB_COLORS, &pvBits, NULL, 0);
+        DeleteDC(hdc);
+    }
+    return hbm;
+}
+
+static HBITMAP
+IntGetWindowBitmap(HWND hwnd, INT cx, INT cy)
+{
+    HBITMAP hbm = NULL;
+    HDC hdc, hdcMem;
+    HGDIOBJ hbmOld;
+
+    hdc = GetWindowDC(hwnd);
+    if (hdc)
+    {
+        hbm = IntCreate32BppBitmap(cx, cy);
+        if (hbm)
+        {
+            hdcMem = CreateCompatibleDC(hdc);
+            if (hdcMem)
+            {
+                hbmOld = SelectObject(hdcMem, hbm);
+                {
+                    BitBlt(hdcMem, 0, 0, cx, cy, hdc, 0, 0, SRCCOPY | CAPTUREBLT);
+                }
+                SelectObject(hdcMem, hbmOld);
+                DeleteDC(hdcMem);
+            }
+        }
+        ReleaseDC(hwnd, hdc);
+    }
+    return hbm;
+}
+
+static VOID
+IntWhiten32BppBitmap(HBITMAP hbm)
+{
+    BITMAP bm;
+    DWORD i, *pdw;
+
+    GetObject(hbm, sizeof(bm), &bm);
+
+    if (bm.bmBitsPixel != 32 || !bm.bmBits)
+        return;
+
+    pdw = (DWORD *)bm.bmBits;
+    for (i = 0; i < bm.bmWidth * bm.bmHeight; ++i)
+    {
+        *pdw = *pdw | 0x00C0C0C0;   // ARGB
+        ++pdw;
+    }
+}
+
+/****************************************************************************/
+
+typedef struct GHOST_DATA
+{
+    HWND hwndTarget;
+    HBITMAP hbm32bpp;
+} GHOST_DATA;
+
+static GHOST_DATA *
+Ghost_GetData(HWND hwnd)
+{
+    if (IsWindowUnicode(hwnd))
+        return (GHOST_DATA *)GetWindowLongPtrW(hwnd, GWLP_USERDATA);
+    else
+        return (GHOST_DATA *)GetWindowLongPtrA(hwnd, GWLP_USERDATA);
+}
+
+static HWND
+Ghost_GetTarget(HWND hwnd)
+{
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+    if (!pData)
+        return NULL;
+    return pData->hwndTarget;
+}
+
+static BOOL
+Ghost_OnCreate(HWND hwnd, CREATESTRUCTW *lpcs)
+{
+    HBITMAP hbm32bpp;
+    HWND hwndTarget, hwndPrev;
+    GHOST_DATA *pData;
+    RECT rc;
+    DWORD style, exstyle;
+    CHAR szTextA[128];
+    WCHAR szTextW[128];
+
+#ifdef __REACTOS__
```

Comment:
I'm not sure what this ifdef expresses. This code doesn't seem to match with any upstream version. Do we diverge from Windows here? If so, a comment describing how/why might be nice.

Reply:
Deleted.